Add logging for parameters

mirromutth opened the pull request:
Motivation:

See also #138 .

When driver executing a text statement (no binding), log will like:

```log
2023-12-31 | 23:59:59.999 | reactor-tcp-nio-1    | DEBUG | i.a.r2dbc.mysql.QUERY | Executing statement [SELECT 1]
```

When user is using client-preparing statements, log will like:

```log
2023-12-31 | 23:59:59.999 | reactor-tcp-nio-1    | DEBUG | i.a.r2dbc.mysql.QUERY | Executing statement [SELECT value FROM test WHERE id=?] with [1]
```

When user is using server-preparing statements, log will like:

```log
2023-12-31 | 23:59:59.999 | reactor-tcp-nio-1    | DEBUG | i.a.r2dbc.mysql.QUERY | Executing statement [INSERT INTO test VALUES(DEFAULT,?)]
2023-12-31 | 23:59:59.999 | reactor-tcp-nio-1    | DEBUG | i.a.r2dbc.mysql.QUERY | Prepared statement 1 [INSERT INTO test VALUES(DEFAULT,?)]
2023-12-31 | 23:59:59.999 | reactor-tcp-nio-1    | DEBUG | i.a.r2dbc.mysql.QUERY | Executing prepared statement 1 with [10]
```

Modification:

- [x] Makes all implementation of `AbstractMySqlParameter` need to override `toString`.
- [x] Add some descriptions of statements logging in README

Result:

More human-friendly SQL debugging logging.

mirromutth committed:
Add logging for parameters

mirromutth committed:
Add a section about statements logging/debugging

jchrys reviewed (commented):
LGTM

However, I'm concerned about the breaking change(`*MySqlParameter#toString` exposes its value, Debug level Query Logger exposes params). Suggest merging with the next minor version update or  log with params only when trace level is enabled. Let's discuss before merging.

mirromutth committed:
Improve doc of custom codec

- Correct parameter interface from `Parameter` to `MySqlParameter`
- Override `toString()` in the implementation of `MySqlParameter`

mirromutth commented:
I fixed a little miss in the README, and I added an override of `toString()` in the example of `MySqlParameter`.

JohnNiang reviewed (approved):
LGTM
